dev/release/utils-create-release-tarball.sh

```diff
@@ -46,18 +52,57 @@ rm -rf ${root_folder}
 # Resolve symbolic and hard links
 rm -rf ${root_folder}.tmp
 mv ${root_folder} ${root_folder}.tmp
-cp -R -L ${root_folder}.tmp ${root_folder}
+cp -a -L ${root_folder}.tmp ${root_folder}
 rm -rf ${root_folder}.tmp
 
-# Create a dummy .git/ directory to download the source files from GitHub with Source Link in C#.
+# Create a dummy .git/ directory to download the source files from
+# GitHub with Source Link in C#. See the followings for more Source
+# Link support and why we need a dummy .git directory:
+# * https://github.com/apache/arrow/issues/21580
+# * https://github.com/apache/arrow/pull/4488
+#
+# We need to set constant timestamp for a dummy .git/ directory for
+# Reproducible Builds. We use mtime of csharp/ for it.
+if stat --help > /dev/null 2>&1; then
+  # GNU stat
+  #
+  # touch accepts YYYYMMDDhhmm.ss format but GNU stat doesn't support
+  # for outputting with the format. So we use date for it.
+  csharp_mtime=$(date --date="$(stat --format="%y" csharp)" +%Y%m%d%H%M.%S)
+else
+  # BSD stat
+  csharp_mtime=$(stat -f %Sm -t %Y%m%d%H%M.%S csharp)
+fi
 dummy_git=${root_folder}/csharp/dummy.git
 mkdir ${dummy_git}
 pushd ${dummy_git}
 echo ${release_hash} > HEAD
 echo "[remote \"origin\"] url = https://github.com/${GITHUB_REPOSITORY:-apache/arrow}.git" >> config
 mkdir objects refs
+find . -exec touch -t "${csharp_mtime}" '{}' ';'
 popd
+touch -t "${csharp_mtime}" ${root_folder}/csharp
 
-# Create new tarball from modified source directory
-tar czf ${tarball} ${root_folder}
+# Create new tarball from modified source directory.
+#
+# We need GNU tar for Reproducible Builds. We want to use the same
```

Comment:
I was hoping we can use something like https://salsa.debian.org/reproducible-builds/strip-nondeterminism but it might not have support for tar files.

Reply:
Wow, I didn't know the tool. It's introduced int https://reproducible-builds.org/docs/stripping-unreproducible-information/ .

The approaches used here are listed in https://reproducible-builds.org/docs/archives/ .

Let's use GNU tar for now because strip-nondeterminism doesn't have tar support.

Reply:
Can we add a comment pointing to  https://reproducible-builds.org/docs/stripping-unreproducible-information/ ?

Reply:
OK. I've added.